miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/RealMiniApp.kt

```diff
@@ -113,11 +138,36 @@ internal class RealMiniApp(
         }
     }
 
+    private suspend fun checkWithManifest(appId: String, versionId: String) {
+        updateManifestFromApi(appId, versionId)
+
+        if (downloadedManifestCache.isRequiredPermissionDenied(
+                getCustomPermissions(appId),
+                versionId
+            )
+        ) throw MiniAppSdkException(ERR_REQUIRED_PERMISSION_DENIED)
+    }
+
+    private suspend fun updateManifestFromApi(appId: String, versionId: String) {
+        val cachedManifest = downloadedManifestCache.readDownloadedManifest(appId, versionId)
+        if (cachedManifest?.versionId != versionId) {
+            val apiManifest = getMiniAppManifest(appId, versionId)
+            downloadedManifestCache.storeDownloadedManifest(appId, CachedManifest(versionId, apiManifest))
+            val customPermissions = miniAppCustomPermissionCache.readPermissions(appId)
+            val manifestPermissions = downloadedManifestCache.getAllPermissions(customPermissions, versionId)
+            miniAppCustomPermissionCache.removePermissionsNotMatching(appId, manifestPermissions)
```

Comment:
It seems this block is unrelated to `updateManifestFromApi`? This is about removing cached permissions - and actually I think we want to do this even if `cachedManifest?.versionId != versionId` is NOT true. This is because the Host App could have set permissions for an already cached mini app.

So maybe you should move this block to `checkWithManifest` just below the call to `updateManifestFromApi`?